### PR TITLE
windows: build natively with msvc compiler

### PR DESCRIPTION
**Pull request by pstorz**
### Thank you for contributing to the Bareos Project!
win-native: changes required to build and test
 
 These changes allow us to build and test Bareos on
 windows natively built with the MSVC compiler.
 
 - Some system tests do not work on windows (yet). These are
   in the TESTS_TO_SKIP list so that there is a central point
   to disable them.
 
 - Required dlls will be copied to the run directory of the produced
   binaries to run system and google tests easier without to hassle too
   much with the PATH settings. (See BareosCopyDllsToBinDir)
 
 - The sysemtests were adapted to be able to support both Debug and
   Release build types. The current value is provided in
   the environment variable CMAKE_CONFIG_TYPE
 
 - Wherever cmake creates files, we now set the NEWLINE_STYLE to UNIX.
   This makes sure that the bash shell that is also used on windows
   to run the system tests gets correct line endings.
 
 - CMake can both produce MS Build Projects for VisualStudio (default)
   and "Ninja Multi-Config" Projects. The latter has the advantage that
   it supports ccache.
 
 - To make ccache work, we need CMake >= 3.25 and CMP0141 set to NEW.
   This way we can set(CMAKE_MSVC_DEBUG_INFORMATION_FORMAT Embedded)
   Which removes compiler switch /Zi which is unupported by ccache.
 
 - Most external dependencies are provided via vcpkg.
 
 - To package the resulting windows binaries, an own nsi installer file
   `winbareos-native.nsi` was created.
 
 - The services used by the systemtests are installed by nssm.
   This is required as it is not possible to start the windows binaries
   detached. Nssm also takes care of stdout and stderr and writes this
   output as logfiles into the systemtests' directory.
 
 - Special start/stop scripts bareos-ctl-* for windows have been added
   which take care of the creation and removeal of nssm services for the
   required binaries.
 
 - The start/stop script `bareos` in the systemtests has been adapted to
   uniformly support multiple directors/filedaemons/storagedaemons.
   To add an additional services, just create a subdirectory
   in the format bareos-<config-name> parallel to the bareos directory in
   the tests. The startscript will automatically create services for
   all directors, filedaemons and storage daemons that have a
   configuration below this directory. For examples check
   `multiple-clients/etc/bareos-secondclient` and
   `autoxflate/etc/bareos-sd1`
 
 - Some scripts remove unexpected windows line ending characters `\r`
   so that the bash shell can handle everything as needed.
 
 - All systemtests have been configured to disable VSS in the filesets
   to speedup testing.
 
 - MSVC Compiler requires the C++ 20 Standard
 
 - Systemtests: Some symlinked files with identical content have been
   split up into multiple files, as on windows the linking from the
   CMAKE_BINARY_DIR to the original files does no work.
 
 - Every ctest call needs to set the configuration, i.e.
   `ctest --build-config Debug/Release ...<other ctest options>`

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)
- [Boy Scout Rule](https://docs.bareos.org/DeveloperGuide/generaldevel.html#boy-scout-rule)

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
~~Check backport line~~
~~Required backport PRs have been created~~

##### Source code quality
- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR

### CODE REVIEW

**Comment by bruno-at-bareos** [comment]
Builds tested correctly on windows 2022, especially upgrade path between older and newer installer.
One defect is bconsole having a weird icon

File: core/src/filed/filed_globals.h [hunk 1]
```diff
@@ -28,16 +28,16 @@ class ConfigurationParser;
 
 namespace filedaemon {
 
-extern ConfigurationParser* my_config;
+BAREOS_IMPORT ConfigurationParser* my_config;
```
Comment (sebsura) [inline core/src/filed/filed_globals.h]: this should be export

File: core/src/lib/output_formatter.cc [hunk 2]
```diff
@@ -1024,7 +1024,14 @@ void OutputFormatter::JsonFinalizeResult(bool result)
       Dmsg0(100, ErrorMsg.c_str());
       JsonSendErrorMessage(ErrorMsg.c_str());
     }
+#  if JANSSON_VERSION_HEX >= 0x020800
+    json_malloc_t my_alloc;
```
Comment (sebsura) [inline core/src/lib/output_formatter.cc]: `my_alloc` is not needed

**Review by sebsura — CHANGES_REQUESTED** [review]
There are some things that need fixing. I already pushed some of the changes i requested here.  Maybe we can go over them together.

The review is only for the first commit for now.  Ill look into the second one now but that should hopefully not contain code changes.

File: systemtests/tests/restore/testrunner-restore-file [hunk 3]
```diff
@@ -34,19 +34,28 @@ export TestName
...
+      rm "$tmp/bareos-restores/$f"
     fi;
 done
+
+if [[ "$(uname -s)" =~ _NT ]]; then
+  filename1="$(cygpath -w -m $file1)"
+  filename2="$(cygpath -w -m $file2)"
+  filename3="$(cygpath -w -m $file3)"
+fi
+
```
Comment (sebsura) [inline systemtests/tests/restore/testrunner-restore-file:58]: where is this used ?

File: systemtests/tests/multiple-clients/test-setup [hunk 4]
```diff
@@ -34,10 +34,10 @@ set -u
 # Fill ${BackupDirectory} with data.
 setup_data
 
+
```
Comment (sebsura) [inline systemtests/tests/multiple-clients/test-setup:37]: why two empty lines ?

File: systemtests/tests/filesets/etc/bareos/command.sh.in [hunk 5]
```diff
@@ -28,36 +28,4 @@ out="@tmpdir@/command-$self.out"
...
-    printf "%s\n" "@tmpdir@/common"
-    printf "%s\n" "@tmpdir@/command-$self"
-    exit 0
-  fi
-  PID=$(ps -o ppid= -p $PID)
-done
-
-echo "EXPECTED PARENT NOT FOUND ($parentbin)" | tee -a $out
-#exit 1
+echo "@tmpdir@/command-$self"
```
Comment (sebsura) [inline systemtests/tests/filesets/etc/bareos/command.sh.in:31]: What did this script do and why did we remove it completely ?

File: systemtests/tests/catalog/testrunner [hunk 6]
```diff
@@ -11,6 +11,8 @@ export TestName
 JobName=backup-bareos-fd
 #shellcheck source=../scripts/functions
 . "${rscripts}"/functions
+
+
```
Comment (sebsura) [inline systemtests/tests/catalog/testrunner:15]: remove

**Review by sebsura — CHANGES_REQUESTED** [review]
The comments for the second commit

File: core/src/include/bareos.h [hunk 7]
```diff
@@ -92,6 +92,7 @@
 #  include <io.h>
 #  include <direct.h>
 #  include <process.h>
+#  define NOMINMAX
 #endif
 
 #if defined(_MSC_VER)
```
[the remark this answers is not in the document]
Comment (sebsura) [reply core/src/include/bareos.h]: maybe im blind

File: core/src/tools/CMakeLists.txt [hunk 8]
```diff
@@ -52,25 +46,32 @@ endif()
...
 if(NOT HAVE_WIN32)
+  add_executable(drivetype drivetype.cc)
+  target_link_libraries(drivetype bareos bareosfind)
+
+  add_executable(fstype fstype.cc)
+  target_link_libraries(fstype bareos bareosfind)
+
+  add_executable(bscrypto bscrypto.cc)
+  target_link_libraries(bscrypto bareos)
+
```
[the remark this answers is not in the document]
Comment (sebsura) [reply core/src/tools/CMakeLists.txt:62]: maybe remove them completely

File: systemtests/scripts/diff.pl.in [hunk 9]
```diff
@@ -16,7 +16,7 @@ use File::Find;
 no warnings 'File::Find';
 use Digest::MD5;
 use Getopt::Long ;
-use Pod::Usage;
+#use Pod::Usage;
```
[the remark this answers is not in the document]
Comment (sebsura) [reply systemtests/scripts/diff.pl.in:19]: only comment out on windows

File: systemtests/scripts/functions [hunk 10]
```diff
@@ -102,8 +112,46 @@ setup_data()
...
    fi
-   $GTAR --directory="$data_dir" -xf "$tarfile" || exit 1
+   if [[ "$(uname -s)" =~ _NT ]]; then
+     $GTAR --directory="$data_dir" -xf "$tarfile" || :  # we expect errors on windows
+   else
+     $GTAR --directory="$data_dir" -xf "$tarfile" || exit 1
+   fi
+  done
+
+# TODO: make this also work on windows
```
[the remark this answers is not in the document]
Comment (sebsura) [reply systemtests/scripts/functions:122]: add better comment

File: systemtests/scripts/setup [hunk 11]
```diff
@@ -1,15 +1,21 @@
...
-"${scripts}/create_bareos_database" "${DBTYPE}" >/dev/null 2>&1
-"${scripts}/make_bareos_tables" "${DBTYPE}" >/dev/null 2>&1
-"${scripts}/grant_bareos_privileges" "${DBTYPE}" >/dev/null 2>&1
+"${scripts}/create_bareos_database"  2>&1 > /dev/null
+
+print_debug "making tables  \"${db_name}\" (${DBTYPE})"
+"${scripts}/make_bareos_tables"  2>&1 > /dev/null
+
+print_debug "granting privileges database \"${db_name}\" (${DBTYPE})"
+"${scripts}/grant_bareos_privileges"   2>&1 > /dev/null || echo "info: ${scripts}/drop_bareos_database failed"
```
[the remark this answers is not in the document]
Comment (sebsura) [reply systemtests/scripts/setup:18]: or update text

File: cmake/BareosSetVariableDefaults.cmake [hunk 12]
```diff
@@ -36,7 +36,75 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "SunOS")
   set(HAVE_EXTENDED_ACL 1)
 endif()
 
-if(USE_RELATIVE_PATHS)
+if(MSVC)
+  set(bindir
+      "C:/Program Files/${CMAKE_PROJECT_NAME}"
```
[the remark this answers is not in the document]
Comment (pstorz) [reply cmake/BareosSetVariableDefaults.cmake:41]: Removed this part for MSVC as it is not needed

File: core/CMakeLists.txt [hunk 13]
```diff
@@ -317,23 +326,29 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
     SYSTEM ${PROJECT_SOURCE_DIR}/src/win32/include
     ${PROJECT_SOURCE_DIR}/src/win32/compat/include
   )
+  if(MSVC)
+    link_libraries(Intl::Intl)
+    include_directories(SYSTEM C:/src/vdi/include SYSTEM C:/src/vss/include)
```
[the remark this answers is not in the document]
Comment (pstorz) [reply core/CMakeLists.txt:331]: Added the headers from microsoft, so we do not need these paths anymore.

File: core/CMakeLists.txt [hunk 14]
```diff
@@ -343,11 +358,10 @@ endif()
...
 
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Werror -Wall -Wextra -Wshadow")
-set(CMAKE_CXX_FLAGS
-    "${CMAKE_CXX_FLAGS} -Werror -Wall -Wextra -Wshadow -Wnon-virtual-dtor"
-)
-
+if(NOT MSVC)
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Werror -Wall -Wextra")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror -Wall -Wextra")
+endif()
```
[the remark this answers is not in the document]
Comment (pstorz) [reply core/CMakeLists.txt:364]: Agreed

File: core/CTestScript.cmake.in [hunk 15]
```diff
@@ -20,8 +20,9 @@ set(CTEST_SOURCE_DIRECTORY @CMAKE_SOURCE_DIR@)
 set(CTEST_BINARY_DIRECTORY @CMAKE_BINARY_DIR@)
 
 ctest_start("Continuous")
-ctest_test(EXCLUDE_LABEL broken
-           PARALLEL_LEVEL 5
-           @REPEAT_UNTIL@
-           )
+ctest_test(
+  EXCLUDE_LABEL broken
+  PARALLEL_LEVEL 10
```
[the remark this answers is not in the document]
Comment (pstorz) [reply core/CTestScript.cmake.in:25]: reverted to original version

File: core/scripts/bareos-ctl-funcs [hunk 16]
```diff
@@ -31,7 +31,7 @@ pidofproc() {
    fi
 
    # Finally try to extract it from ps
-   pid=`${PSCMD} | grep -m 1 $base | ${AWK} '{ print $1 }' | tr '\n' ' '`
+   pid=$(ps | grep ${base} | awk '{print $1}' | head -n1)
```
[the remark this answers is not in the document]
Comment (pstorz) [reply core/scripts/bareos-ctl-funcs:34]: I went back to the original Version.

File: core/src/include/bareos.h [hunk 17]
```diff
@@ -92,6 +92,7 @@
 #  include <io.h>
 #  include <direct.h>
 #  include <process.h>
+#  define NOMINMAX
```
[the remark this answers is not in the document]
Comment (pstorz) [reply core/src/include/bareos.h:95]: Done

File: core/src/lib/message.h [hunk 18]
```diff
@@ -5,20 +5,20 @@
...
+
+ This program is distributed in the hope that it will be useful, but
+ WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with this program; if not, write to the Free Software
+ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ 02110-1301, USA.
```
[the remark this answers is not in the document]
Comment (pstorz) [reply core/src/lib/message.h:21]: Done

File: core/src/win32/compat/glob.cc [hunk 19]
```diff
@@ -674,8 +731,8 @@ static int glob_match(const char* pattern,
 
   /* Begin by separating out any path prefix from the glob pattern.
    */
-  std::string dirbuf(pattern);
-  const char* dir = dirname(dirbuf.data());
+  char dirbuf[MAX_PATH];
+  const char* dir = dirname((char*)memcpy(dirbuf, pattern, sizeof(dirbuf)));
```
[the remark this answers is not in the document]
Comment (pstorz) [reply core/src/win32/compat/glob.cc:735]: I guess this needs to be verified again. The master code does not build becaues of: `error C3863: array type 'char [2+dirlen+matchlen]' is not assignable`

File: systemtests/CMakeLists.txt [hunk 20]
```diff
@@ -19,8 +19,65 @@
...
+      system:virtualfull-bscan
+  )
+else()
+  set(DISABLED_TESTS)
+endif()
+
 project(bareos-systemtests)
 
+if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
+  if(MSVC)
```
[the remark this answers is not in the document]
Comment (pstorz) [reply systemtests/CMakeLists.txt:76]: As we only run the systemtests on windows native, you are right that we can make this simpler.

File: systemtests/CMakeLists.txt [hunk 21]
```diff
@@ -200,13 +260,38 @@ if(MINIO)
...
 endif()
 
-configure_file(
-  "CTestCustom.cmake.in" "${CMAKE_BINARY_DIR}/CTestCustom.cmake" @ONLY
-)
-configure_file(
-  "ctest_custom_pretest.sh.in"
-  "${CMAKE_CURRENT_BINARY_DIR}/ctest_custom_pretest.sh" @ONLY
-)
+if(${CMAKE_SYSTEM_NAME} MATCHES "Windows" AND MSVC)
```
[the remark this answers is not in the document]
Comment (pstorz) [reply systemtests/CMakeLists.txt:263]: Makes sense

File: systemtests/environment.in [hunk 22]
```diff
@@ -1,3 +1,7 @@
+# CMAKE_CONFIG_TYPE is required on windows, on other OS it needs to be empty
+if [[ ! "$(uname -s)" =~ _NT ]]; then
+  CMAKE_CONFIG_TYPE=""
+fi
```
[the remark this answers is not in the document]
Comment (pstorz) [reply systemtests/environment.in]: Done

File: systemtests/environment.in [hunk 23]
```diff
@@ -109,16 +116,31 @@ export BAREOS_WEBUI_LOG_PATH=${logdir}
...
+../../../core/src/console/${CMAKE_CONFIG_TYPE}:\
+../../../core/src/dird/${CMAKE_CONFIG_TYPE}:\
+../../../core/src/filed/${CMAKE_CONFIG_TYPE}:\
+../../../core/src/stored/${CMAKE_CONFIG_TYPE}:\
+../../../core/src/tests/${CMAKE_CONFIG_TYPE}:\
+../../../core/src/tools/${CMAKE_CONFIG_TYPE}:\
+../../../core/src/plugins/filed/python/python3modules/${CMAKE_CONFIG_TYPE}:\
+../../../core/src/plugins/stored/python/python3modules/${CMAKE_CONFIG_TYPE}:\
+../../../core/src/plugins/dird/python/python3modules/${CMAKE_CONFIG_TYPE}:\
+$PATH
```
[the remark this answers is not in the document]
Comment (pstorz) [reply systemtests/environment.in:138]: done

File: systemtests/scripts/diff.pl.in (hunk 9 above)
[the remark this answers is not in the document]
Comment (pstorz) [reply systemtests/scripts/diff.pl.in:19]: Done

File: systemtests/scripts/functions (hunk 10 above)
[the remark this answers is not in the document]
Comment (pstorz) [reply systemtests/scripts/functions:122]: done

File: systemtests/scripts/functions [hunk 24]
```diff
@@ -102,8 +112,46 @@ setup_data()
...
+ if [[ "$(uname -s)" =~ _NT ]]; then
+   pushd "$data_dir" || echo "pushd failed"
```
[the remark this answers is not in the document]
Comment (pstorz) [reply systemtests/scripts/functions:124]: agreed

File: systemtests/scripts/functions [hunk 25]
```diff
@@ -102,8 +112,46 @@ setup_data()
...
+    weird-files/soft-file2 \
+    weird-files/softlink-to-softlinkee \
+    weird-files/a-file-with* \
+    weird-files/normalfile \
+    weird-files/another-hardlink-to-normalfile \
+    weird-files/hardlink-to-normalfile \
+
+  do rm -Rvf "$f" || echo "failed to remove \"$f\""
   done
+   popd || echo "popd failed"
```
[the remark this answers is not in the document]
Comment (pstorz) [reply systemtests/scripts/functions:153]: done

File: systemtests/scripts/functions [hunk 26]
```diff
@@ -428,23 +476,18 @@ check_duration_lt()
...
 }
 
 start_sd() {
-   "${BAREOS_STORAGEDAEMON_BINARY}" -c "${BAREOS_CONFIG_DIR}" -t </dev/null
-   "${BAREOS_STORAGEDAEMON_BINARY}" -c "${BAREOS_CONFIG_DIR}" -f </dev/null &
-   sd_pid=$!
+ ${rscripts}/bareos-ctl-sd start
 }
 
+
```
[the remark this answers is not in the document]
Comment (pstorz) [reply systemtests/scripts/functions:490]: done

File: systemtests/scripts/functions [hunk 27]
```diff
@@ -842,8 +876,8 @@ check_restore_diff()
    result=0
 
    restorepath="$restoredirectory/${dest}"
-
-   if ! "$rscripts/diff.pl" -d "${restorepath}" -s "${dest}"; then
+   restorepath=$(sed 's#/\(.\):#/\1#g'<<< $restorepath)
```
[the remark this answers is not in the document]
Comment (pstorz) [reply systemtests/scripts/functions]: agree

File: systemtests/scripts/functions [hunk 28]
```diff
@@ -1190,7 +1237,7 @@ cwd="$(pwd)"
 
 # Source the configuration variables
 # shellcheck source=../environment.in
-. "${cwd}/environment"
+#. "${cwd}/environment"
```
[the remark this answers is not in the document]
Comment (pstorz) [reply systemtests/scripts/functions:1249]: removed

File: systemtests/tests/bareos-basic/testrunner-status-subscriptions [hunk 29]
```diff
@@ -40,7 +40,7 @@ run_bconsole
 sed -i'.bak' -e '/^@/d' -e '/^Bareos version:/d' -e '/^Binary info:/d' -e '/^Report time:/d' -e '/^Checksum:/d' -e "/^Estimate only./d" "$tmp"/status-subscription*.txt
 
 for f in status-subscriptions status-subscriptions-detail status-subscriptions-unknown; do
-  if ! diff -u expected/$f.txt "$tmp/$f.txt"; then
+  if ! diff -w -u expected/$f.txt "$tmp/$f.txt"; then
```
[the remark this answers is not in the document]
Comment (pstorz) [reply systemtests/tests/bareos-basic/testrunner-status-subscriptions:43]: changed to use long options: --ignore-all-space --unified

File: systemtests/tests/restore/testrunner-restore-file (hunk 3 above)
Comment (pstorz) [reply systemtests/tests/restore/testrunner-restore-file:58]: done

File: systemtests/tests/multiple-clients/test-setup (hunk 4 above)
Comment (pstorz) [reply systemtests/tests/multiple-clients/test-setup:37]: done

File: systemtests/tests/filesets/etc/bareos/command.sh.in (hunk 5 above)
Comment (pstorz) [reply systemtests/tests/filesets/etc/bareos/command.sh.in:31]: done

**Review by sebsura — APPROVED** [review]
Thanks for the great work!

**Pull request merged**